File: android/app/src/main/java/com/summit/summitberber/utils/PdfGenerator.kt
package com.summit.summitberber.utils

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Typeface
import android.graphics.pdf.PdfDocument
import com.summit.summitberber.data.Report
import java.io.File
import java.io.FileOutputStream
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

object PdfGenerator {

    fun generateReportPdf(context: Context, report: Report): File? {
        val pdfDocument = PdfDocument()
        
        // A4 Dimensions: 595 x 842 points
        val pageInfo = PdfDocument.PageInfo.Builder(595, 842, 1).create()
        val page = pdfDocument.startPage(pageInfo)
        
        val canvas: Canvas = page.canvas
        val paint = Paint()
        val textPaint = Paint()
        
        // Background
        canvas.drawColor(Color.parseColor("#F7FBF0")) // Muted off-white
        
        // Header
        paint.color = Color.parseColor("#2E7D32") // Forest Green
        canvas.drawRect(0f, 0f, 595f, 80f, paint)
        
        // Header Text
        textPaint.color = Color.WHITE
        textPaint.textSize = 22f
        textPaint.typeface = Typeface.create(Typeface.DEFAULT, Typeface.BOLD)
        canvas.drawText("SUMMITBERBER - TRAIL REPORT", 30f, 48f, textPaint)
        
        // Status tag (Draft/Completed)
        paint.color = if (report.isDraft) Color.parseColor("#E65100") else Color.parseColor("#2E7D32")
        val statusText = if (report.isDraft) "DRAFT" else "COMPLETED"
        canvas.drawRoundRect(460f, 25f, 565f, 55f, 8f, 8f, paint)
        
        textPaint.color = Color.WHITE
        textPaint.textSize = 10f
        textPaint.typeface = Typeface.create(Typeface.DEFAULT, Typeface.BOLD)
        textPaint.textAlign = Paint.Align.CENTER
        canvas.drawText(statusText, 512.5f, 43f, textPaint)
        textPaint.textAlign = Paint.Align.LEFT // reset
        
        // Metadata / Main info
        textPaint.color = Color.parseColor("#1C1D1A")
        textPaint.textSize = 18f
        textPaint.typeface = Typeface.create(Typeface.DEFAULT, Typeface.BOLD)
        canvas.drawText(report.title, 30f, 120f, textPaint)
        
        textPaint.color = Color.parseColor("#5C5E57")
        textPaint.textSize = 12f
        textPaint.typeface = Typeface.create(Typeface.DEFAULT, Typeface.ITALIC)
        canvas.drawText("Region / Park: ${report.park}", 30f, 140f, textPaint)
        
        val sdf = SimpleDateFormat("yyyy-MM-dd HH:mm", Locale.getDefault())
        canvas.drawText("Date: ${sdf.format(Date(report.timestamp))}", 30f, 155f, textPaint)
        
        // Divider
        paint.color = Color.parseColor("#E1E4D5")
        canvas.drawRect(30f, 170f, 565f, 172f, paint)
        
        // METRICS SECTION (Table style)
        textPaint.color = Color.parseColor("#2E7D32")
        textPaint.textSize = 14f
        textPaint.typeface = Typeface.create(Typeface.DEFAULT, Typeface.BOLD)
        canvas.drawText("TRAIL METRICS", 30f, 200f, textPaint)
        
        // Draw standard Table boxes
        paint.color = Color.parseColor("#FFFFFF")
        canvas.drawRoundRect(30f, 215f, 565f, 275f, 8f, 8f, paint)
        paint.style = Paint.Style.STROKE
        paint.color = Color.parseColor("#C3C8BB")
        paint.strokeWidth = 1f
        canvas.drawRoundRect(30f, 215f, 565f, 275f, 8f, 8f, paint)
        paint.style = Paint.Style.FILL // reset
        
        // Grid vertical dividers
        canvas.drawLine(208.3f, 215f, 208.3f, 275f, paint)
        canvas.drawLine(386.6f, 215f, 386.6f, 275f, paint)
        
        // Fill Table Titles
        textPaint.color = Color.parseColor("#5C5E57")
        textPaint.textSize = 11f
        textPaint.typeface = Typeface.create(Typeface.DEFAULT, Typeface.NORMAL)
        canvas.drawText("Distance", 60f, 235f, textPaint)
        canvas.drawText("Duration", 238.3f, 235f, textPaint)
        canvas.drawText("Elevation Gain", 416.6f, 235f, textPaint)
        
        // Fill Table Values
        textPaint.color = Color.parseColor("#1C1D1A")
        textPaint.textSize = 16f
        textPaint.typeface = Typeface.create(Typeface.DEFAULT, Typeface.BOLD)
        canvas.drawText("${report.distance} km", 60f, 260f, textPaint)
        canvas.drawText(report.duration, 238.3f, 260f, textPaint)
        canvas.drawText("${report.elevationGain} m", 416.6f, 260f, textPaint)
        
        // DESCRIPTION
        textPaint.color = Color.parseColor("#2E7D32")
        textPaint.textSize = 14f
        canvas.drawText("DESCRIPTION", 30f, 310f, textPaint)
        
        textPaint.color = Color.parseColor("#1C1D1A")
        textPaint.textSize = 11f
        textPaint.typeface = Typeface.create(Typeface.DEFAULT, Typeface.NORMAL)
        
        // Simple word wrap
        var y = 335f
        val words = report.description.split(" ")
        var line = ""
        for (word in words) {
            val testLine = if (line.isEmpty()) word else "$line $word"
            val width = textPaint.measureText(testLine)
            if (width > 535f) {
                canvas.drawText(line, 30f, y, textPaint)
                y += 18f
                line = word
            } else {
                line = testLine
            }
        }
        if (line.isNotEmpty()) {
            canvas.drawText(line, 30f, y, textPaint)
            y += 25f
        }
        
        // GPX LINK
        if (!report.gpxPath.isNullOrEmpty()) {
            paint.color = Color.parseColor("#EBF3E7")
            canvas.drawRoundRect(30f, y, 565f, y + 35f, 6f, 6f, paint)
            
            textPaint.color = Color.parseColor("#2E7D32")
            textPaint.textSize = 10f
            textPaint.typeface = Typeface.create(Typeface.DEFAULT, Typeface.BOLD)
            canvas.drawText("ATTACHED GPX ROUTE DATA", 45f, y + 16f, textPaint)
            
            textPaint.color = Color.parseColor("#5C5E57")
            textPaint.typeface = Typeface.create(Typeface.DEFAULT, Typeface.NORMAL)
            val fileName = report.gpxPath.substringAfterLast("/")
            canvas.drawText("Linked file: $fileName", 45f, y + 28f, textPaint)
            
            y += 55f
        }
        
        // SIGNATURE / FOOTER at bottom
        textPaint.color = Color.parseColor("#8E9182")
        textPaint.textSize = 10f
        textPaint.typeface = Typeface.create(Typeface.DEFAULT, Typeface.NORMAL)
        canvas.drawText("Generated offline via SummitBerber V1 Mountain Productivity Suite.", 30f, 800f, textPaint)
        canvas.drawText("Page 1 of 1", 515f, 800f, textPaint)
        
        pdfDocument.finishPage(page)
        
        // Write to file
        val reportsDir = File(context.getExternalFilesDir(null), "Reports")
        if (!reportsDir.exists()) {
            reportsDir.mkdirs()
        }
        val cleanTitle = report.title.replace("\\s+".toRegex(), "_")
        val file = File(reportsDir, "Report_${cleanTitle}_${report.timestamp}.pdf")
        
        return try {
            pdfDocument.writeTo(FileOutputStream(file))
            pdfDocument.close()
            file
        } catch (e: Exception) {
            e.printStackTrace()
            pdfDocument.close()
            null
        }
    }
}

File: src/utils/pdfGenerator.ts
import { jsPDF } from "jspdf";
import { TrailReport } from "../types";

export function generateReportWebPdf(report: TrailReport) {
  const doc = new jsPDF({
    orientation: "portrait",
    unit: "mm",
    format: "a4"
  });
  
  // A4 size is 210 x 297 mm
  
  // Banner background (Forest Green #2E7D32)
  doc.setFillColor(46, 125, 50);
  doc.rect(0, 0, 210, 30, "F");
  
  // Title
  doc.setTextColor(255, 255, 255);
  doc.setFont("helvetica", "bold");
  doc.setFontSize(15);
  doc.text("SUMMITBERBER - TRAIL ASSESSMENT REPORT", 12, 18);
  
  // Status tag
  doc.setFillColor(report.isDraft ? 230 : 46, report.isDraft ? 81 : 125, report.isDraft ? 0 : 50);
  doc.roundedRect(172, 9, 26, 10, 2, 2, "F");
  doc.setTextColor(255, 255, 255);
  doc.setFontSize(8.5);
  doc.setFont("helvetica", "bold");
  doc.text(report.isDraft ? "DRAFT" : "FINALIZED", 185, 15.5, { align: "center" });
  
  // Main metadata
  doc.setTextColor(28, 29, 26);
  doc.setFont("helvetica", "bold");
  doc.setFontSize(14);
  doc.text(report.title || "Unnamed Assessment", 12, 42);
  
  doc.setTextColor(92, 94, 87);
  doc.setFont("helvetica", "italic");
  doc.setFontSize(9.5);
  doc.text(`Region / Park: ${report.park || "Not Specified"}`, 12, 48);
  doc.text(`Date Filed: ${new Date(report.timestamp).toLocaleString()}`, 12, 53);
  
  // Divider line
  doc.setDrawColor(225, 228, 213);
  doc.setLineWidth(0.4);
  doc.line(12, 57, 198, 57);
  
  // Metrics Grid table
  doc.setTextColor(46, 125, 50);
  doc.setFont("helvetica", "bold");
  doc.setFontSize(10.5);
  doc.text("TRAIL METRICS", 12, 65);
  
  // Outer border of table
  doc.setDrawColor(195, 200, 187);
  doc.setFillColor(255, 255, 255);
  doc.roundedRect(12, 69, 186, 18, 2, 2, "FD");
  
  // Column dividers
  doc.line(74, 69, 74, 87);
  doc.line(136, 69, 136, 87);
  
  // Column Labels
  doc.setTextColor(92, 94, 87);
  doc.setFontSize(8);
  doc.setFont("helvetica", "normal");
  doc.text("Total Distance", 20, 74);
  doc.text("Est. Duration", 82, 74);
  doc.text("Elevation Gain", 144, 74);
  
  // Column values
  doc.setTextColor(28, 29, 26);
  doc.setFontSize(10.5);
  doc.setFont("helvetica", "bold");
  doc.text(`${report.distance} km`, 20, 81);
  doc.text(report.duration || "00:00", 82, 81);
  doc.text(`${report.elevationGain} m`, 144, 81);
  
  // Description section
  doc.setTextColor(46, 125, 50);
  doc.setFont("helvetica", "bold");
  doc.setFontSize(10.5);
  doc.text("DESCRIPTION & SAFETY ANALYSIS", 12, 97);
  
  doc.setTextColor(28, 29, 26);
  doc.setFont("helvetica", "normal");
  doc.setFontSize(9.5);
  
  // Split description text to fit PDF width
  const splitText = doc.splitTextToSize(report.description || "No description provided.", 186);
  doc.text(splitText, 12, 103);
  
  let currentY = 103 + (splitText.length * 4.8) + 8;
  
  // GPX Track details if attached
  if (report.gpxFileName) {
    doc.setFillColor(235, 243, 231);
    doc.roundedRect(12, currentY, 186, 12, 1.5, 1.5, "F");
    
    doc.setTextColor(46, 125, 50);
    doc.setFont("helvetica", "bold");
    doc.setFontSize(8);
    doc.text("GPX ROUTE ATTACHED", 16, currentY + 5);
    
    doc.setTextColor(92, 94, 87);
    doc.setFont("helvetica", "normal");
    doc.text(`File: ${report.gpxFileName}`, 16, currentY + 9);
    
    currentY += 18;
  }
  
  // Photo gallery section
  if (report.images && report.images.length > 0) {
    doc.setTextColor(46, 125, 50);
    doc.setFont("helvetica", "bold");
    doc.setFontSize(10.5);
    doc.text("FIELD DOCUMENTATION PHOTOS", 12, currentY);
    currentY += 6;
    
    // Draw photos grid
    report.images.forEach((img, idx) => {
      if (currentY + 48 > 280) {
        doc.addPage();
        currentY = 20;
      }
      
      const col = idx % 2;
      const row = Math.floor(idx / 2);
      
      const x = col === 0 ? 12 : 110;
      const y = currentY + (row * 46);
      
      try {
        // Draw image
        doc.addImage(img.dataUrl, "JPEG", x, y, 88, 36);
        doc.setFillColor(242, 245, 238);
        doc.rect(x, y + 36, 88, 5, "F");
        doc.setTextColor(50, 50, 50);
        doc.setFontSize(7.5);
        doc.setFont("helvetica", "normal");
        doc.text(img.caption || `Photo #${idx + 1}`, x + 3, y + 39.5);
      } catch (e) {
        // Fallback placeholder box
        doc.setDrawColor(200, 200, 200);
        doc.rect(x, y, 88, 36, "S");
        doc.setTextColor(150, 150, 150);
        doc.text("[Image render placeholder]", x + 15, y + 18);
      }
    });
  }
  
  // Footer
  doc.setTextColor(142, 145, 130);
  doc.setFontSize(8);
  doc.setFont("helvetica", "normal");
  doc.text("Generated via SummitBerber V1 Mountain Productivity Suite", 12, 288);
  doc.text("Page 1 of 1", 198, 288, { align: "right" });
  
  // Save
  const cleanTitle = (report.title || "Report").replace(/\s+/g, "_");
  doc.save(`SummitBerber_Report_${cleanTitle}.pdf`);
}
